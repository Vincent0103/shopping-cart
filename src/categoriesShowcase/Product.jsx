const Product = ({
  imgSrcs,
  alt,
  text,
  additionalImgStyles,
  additionalContainerStyles,
  imgDirection,
}) => {
  const Picture = () => (
    <picture className={`${additionalImgStyles} pointer-events-none`}>
      <source srcSet={imgSrcs[0]} type="image/webp" />
      <img className="object-cover" src={imgSrcs[1]} alt="" />
    </picture>
  );

  const H3 = () => (
    <h3
      className={
        "text-9xl max-xl:text-7xl max-md:text-5xl max-sm:text-3xl font-black font-jost tracking-wide"
      }
    >
      {text}
    </h3>
  );

  return (
    <div
      className={`relative flex justify-center items-center max-lg:gap-2 ${
        imgDirection ? "w-full" : ""
      } ${
        additionalContainerStyles
          ? `${additionalContainerStyles} mb-32 max-xl:mb-16 max-md:mb-10 max-sm:mb-8`
          : "my-32 max-xl:my-16 max-md:my-10 max-sm:my-8"
      }
      `}
    >
      {imgDirection === "right" ? (
        <>
          <H3 />
          <Picture />
        </>
      ) : (
        <>
          <Picture />
          <H3 />
        </>
      )}
    </div>
  );
};

export default Product;
