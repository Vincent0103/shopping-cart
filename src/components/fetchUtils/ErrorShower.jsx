import PropTypes from "prop-types";

const ErrorShower = ({ errorMsg }) => (
  <div className="size-full bg-transparent backdrop-blur-xl">
    <div className="flex h-[80vh] w-full items-center justify-center">
      <h1 className="text-7xl font-semibold text-red-400">Error: {errorMsg}</h1>
    </div>
  </div>
);

ErrorShower.propTypes = {
  errorMsg: PropTypes.string.isRequired,
};

export default ErrorShower;
