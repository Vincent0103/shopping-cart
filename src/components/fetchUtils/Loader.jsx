const Loader = () => (
  <div className="size-full bg-transparent backdrop-blur-xl">
    <div className="flex h-[80vh] w-full items-center justify-center">
      <h1 className="text-[10vw] font-black">LOADING...</h1>
    </div>
  </div>
);

export default Loader;
