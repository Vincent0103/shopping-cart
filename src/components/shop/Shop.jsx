import { Link, useParams } from "react-router-dom";
import Category from "./Category";
import PropTypes from "prop-types";

const Li = ({ text, linkTo, isSelected = false }) => (
  <Link className="flex size-full items-center justify-center" to={linkTo}>
    <li
      className={`flex size-full items-center justify-end rounded-md px-4 py-2 transition-colors duration-100 max-md:px-2 max-md:py-1
          ${isSelected ? "bg-gradient-to-r from-accent-200 to-accent-400 text-text-950 shadow-lg" : "hover:bg-accent-200/50 hover:text-text-950"}`}
    >
      <h3>{text}</h3>
    </li>
  </Link>
);

const Shop = () => {
  let { name } = useParams();

  const validCategories = [
    "electronics",
    "jewelry",
    "mens-clothing",
    "womens-clothing",
  ];

  name = name.split("/")[0];
  if (!validCategories.includes(name)) name = "";

  return (
    <div className="size-full">
      <div className="relative top-20 size-full">
        <nav
          aria-label="Category"
          className="fixed top-20 w-[430px] p-3 max-xl:w-[350px] max-md:w-[200px] max-md:p-1.5"
        >
          <div className="min-h-calc-100vh-96px w-full rounded-xl bg-accent-900 shadow-xl">
            <div className="flex justify-end rounded-xl bg-accent-500 py-4 pr-6 shadow-xl max-md:rounded-md max-md:py-2 max-md:pr-3">
              <h3 className="text-2xl font-bold max-md:text-lg">Category</h3>
            </div>
            <ul
              className="flex flex-col items-center justify-center gap-2 p-1 py-3 text-2xl font-semibold max-md:py-1 
              max-md:text-base max-md:font-medium"
            >
              <Li text={"All"} linkTo={"/shop"} isSelected={name === ""} />
              <Li
                text={"Electronics"}
                linkTo={"/shop/electronics"}
                isSelected={name === "electronics"}
              />
              <Li
                text={"Jewelry"}
                linkTo={"/shop/jewelry"}
                isSelected={name === "jewelry"}
              />
              <Li
                text={"Men's Clothing"}
                linkTo={"/shop/mens-clothing"}
                isSelected={name === "mens-clothing"}
              />
              <Li
                text={"Women's Clothing"}
                linkTo={"/shop/womens-clothing"}
                isSelected={name === "womens-clothing"}
              />
            </ul>
          </div>
        </nav>
        <section
          className="relative left-[430px] flex w-full-minus-left-430px flex-wrap justify-center gap-8 p-3 max-xl:left-[350px]
          max-xl:w-full-minus-left-350px max-md:left-[200px] max-md:w-full-minus-left-200px"
        >
          <Category categoryName={name} />
        </section>
      </div>
    </div>
  );
};

Li.propTypes = {
  text: PropTypes.string.isRequired,
  linkTo: PropTypes.string.isRequired,
  isSelected: PropTypes.bool,
};

export default Shop;
