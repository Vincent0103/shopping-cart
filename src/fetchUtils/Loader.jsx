const Loader = () => (
  <div className="size-full bg-transparent backdrop-blur-xl">
    <div className="w-full h-[80vh] flex items-center justify-center">
      <h1 className="rotate-1 text-[10vw] font-black">LOADING...</h1>
    </div>
  </div>
);

export default Loader;
