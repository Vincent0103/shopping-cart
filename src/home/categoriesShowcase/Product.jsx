import PropTypes from "prop-types";

const Product = ({
  imgSrcs,
  alt,
  text,
  imgDirection,
  additionalImgStyles = "",
  additionalContainerStyles = "",
}) => {
  const Picture = () => (
    <picture
      className={`${additionalImgStyles} pointer-events-none z-[1] select-none transition-transform group-hover:scale-105`}
    >
      <source srcSet={imgSrcs[0]} type="image/webp" />
      <img className="object-cover" src={imgSrcs[1]} alt={alt} />
    </picture>
  );

  const H3 = () => (
    <h3
      className={
        "font-jost text-9xl font-black tracking-wide max-xl:text-7xl max-md:text-5xl max-sm:text-3xl"
      }
    >
      {text}
    </h3>
  );

  return (
    <button
      type="button"
      className={`relative flex items-center justify-center max-lg:gap-2 ${
        imgDirection ? "w-full" : ""
      } ${
        additionalContainerStyles
          ? `${additionalContainerStyles} mb-32 max-xl:mb-16 max-md:mb-10 max-sm:mb-8`
          : "my-24 max-xl:my-12 max-md:my-8 max-sm:my-4"
      }
      group`}
    >
      {imgDirection === "right" ? (
        <>
          <H3 />
          <Picture />
        </>
      ) : (
        <>
          <Picture />
          <H3 />
        </>
      )}
    </button>
  );
};

Product.propTypes = {
  imgSrcs: PropTypes.arrayOf(PropTypes.string).isRequired,
  alt: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  imgDirection: PropTypes.string,
  additionalImgStyles: PropTypes.string,
  additionalContainerStyles: PropTypes.string,
};

export default Product;
