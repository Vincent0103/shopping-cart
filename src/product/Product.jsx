import { useContext, useState } from "react";
import { Link, useLocation } from "react-router-dom";
import { produce } from "immer";
import { v4 as uuid } from "uuid";
import { Minus, Plus, ShoppingCart } from "lucide-react";
import BgPurpleNoise from "../assets/purpleNoise.jpg";
import { AppContext } from "../AppContext.jsx";
import { toTitle } from "../utils.js";

const Product = () => {
  const location = useLocation();
  const { setPopupCartState, setCart } = useContext(AppContext);

  const { id, imgSrc, alt, title, desc, price, productCategory } = location.state;
  const [productAmount, setProductAmount] = useState(1);

  const handlePlusOrMinusClick = (action) => {
    if (action === "add") setProductAmount((prev) => prev + 1);
    else if (productAmount - 1 > 0) setProductAmount((prev) => prev - 1);
  };

  const handleAddToCartClick = () => {
    setPopupCartState((prev) => !prev);
    setCart(
      produce((draft) => {

        draft.unshift({
          id,
          imgSrc,
          alt,
          title,
          productAmount,
        });
      }),
    );
  };

  return (
    <div
      style={{ backgroundImage: `url(${BgPurpleNoise})` }}
      className="flex h-screen min-h-screen w-full items-center justify-center font-extrabold text-text-50"
    >
      <section className="flex h-3/5 w-full max-w-screen-2xl flex-col items-center justify-center gap-10">
        <nav className="self-start" aria-label="Previous links">
          <ul className="flex text-lg font-light text-zinc-500">
            <Link to={`/shop/${productCategory}`}>
              <li className="underline">{toTitle(productCategory) || "All"}</li>
            </Link>
            <p>&nbsp;&gt;&nbsp;</p>
            <li className="max-w-60 overflow-hidden text-ellipsis text-nowrap">
              {title}
            </li>
          </ul>
        </nav>
        <div className="flex size-full gap-20">
          <div className="flex h-full w-1/2 justify-end">
            <img
              className="h-full rounded-sm object-cover"
              src={imgSrc}
              alt={alt}
            />
          </div>
          <div className="flex w-1/2 flex-col gap-6">
            <div className="flex flex-col items-start gap-3">
              <h1 className="h-[66px] w-full overflow-hidden text-ellipsis text-nowrap text-6xl">
                {title}
              </h1>
              <h2 className="text-3xl font-semibold">{price}</h2>
            </div>
            <hr className="bg-white" />
            <div>
              <p className="text-2xl font-normal leading-normal">{desc}</p>
            </div>
            <div className="flex items-center gap-2">
              <Plus
                className="cursor-pointer"
                size={40}
                color="#ececf8"
                onClick={() => handlePlusOrMinusClick("add")}
              />
              <input
                className="flex size-12 items-center justify-center
              rounded-full border-2 border-accent-500 bg-background-100
              text-center text-2xl text-background-900"
                type="text"
                name=""
                id=""
                value={productAmount}
              />
              <Minus
                className="cursor-pointer"
                size={40}
                color="#ececf8"
                onClick={() => handlePlusOrMinusClick("remove")}
              />
            </div>
            <button
              onClick={handleAddToCartClick}
              className="shadow-accent-btn flex h-14 items-center
              gap-4 rounded-md bg-primary-500 px-4 text-2xl
              font-medium shadow-primary-500 transition-shadow
              duration-100"
              type="submit"
            >
              <ShoppingCart size={32} />
              <p>Add to Cart</p>
            </button>
          </div>
        </div>
      </section>
    </div>
  );
};

export default Product;
