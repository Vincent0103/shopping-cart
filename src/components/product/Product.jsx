import BgPurpleNoise from "../../assets/purpleNoise.jpg";
import { Minus, Plus, ShoppingCart } from "lucide-react";
import { useContext, useState } from "react";
import { Link, useParams } from "react-router-dom";
import { produce } from "immer";
import { AppContext } from "../../AppContext.jsx";
import { toTitle } from "../../utils.js";
import Loader from "../fetchUtils/Loader.jsx";
import ErrorShower from "../fetchUtils/ErrorShower.jsx";
import useData from "../fetchUtils/Fetch.jsx";

const Product = () => {
  const { setPopupCartState, setCart } = useContext(AppContext);

  const { id } = useParams();
  const [productAmount, setProductAmount] = useState(1);
  const {
    data: productInfos,
    errorMessage,
    isLoading,
  } = useData(`https://fakestoreapi.com/products/${id}`, [id]);

  const handlePlusOrMinusClick = (action) => {
    if (action === "add") setProductAmount((prev) => prev + 1);
    else if (productAmount - 1 > 0) setProductAmount((prev) => prev - 1);
  };

  const handleAddToCartClick = () => {
    setPopupCartState(true);
    setCart(
      produce((draft) => {
        const itemIndex = draft.findIndex(({ id: draftId }) => draftId === id);
        if (itemIndex !== -1) draft[itemIndex].productAmount += productAmount;
        else {
          const { image, title, description, price, category } = productInfos;
          draft.unshift({
            imgSrc: image,
            desc: description,
            productCategory: category,
            id,
            title,
            price: Math.round(price * 100) / 100,
            productAmount,
          });
        }
      }),
    );
  };

  if (isLoading) return <Loader />;
  else if (errorMessage) return <ErrorShower errorMsg={errorMessage} />;
  else if (productInfos) {
    return (
      <div
        style={{ backgroundImage: `url(${BgPurpleNoise})` }}
        className="flex h-screen min-h-screen w-full items-center justify-center font-extrabold text-text-50"
      >
        <section className="flex h-3/5 w-full max-w-screen-2xl flex-col items-center justify-center gap-10">
          <nav className="self-start" aria-label="Previous links">
            <ul className="flex text-lg font-light text-zinc-500">
              <Link to={`/shop/${productInfos.category}`}>
                <li className="underline">
                  {toTitle(productInfos.category) || "All"}
                </li>
              </Link>
              <p>&nbsp;&gt;&nbsp;</p>
              <li className="max-w-60 overflow-hidden text-ellipsis text-nowrap">
                {productInfos.title}
              </li>
            </ul>
          </nav>
          <div className="flex size-full gap-20">
            <div className="flex h-full w-1/2 justify-end">
              <img
                className="h-full rounded-sm object-cover"
                src={productInfos.image}
                alt={`Product: ${productInfos.title}`}
              />
            </div>
            <div className="flex w-1/2 flex-col gap-6">
              <div className="flex flex-col items-start gap-3">
                <h1 className="h-[66px] w-full overflow-hidden text-ellipsis text-nowrap font-jost text-6xl">
                  {productInfos.title}
                </h1>
                <h2 className="text-2xl font-semibold">
                  {productInfos.price}$
                </h2>
              </div>
              <hr className="bg-white" />
              <div>
                <p className="text-2xl font-normal leading-normal">
                  {productInfos.description}
                </p>
              </div>
              <div className="flex items-center gap-2">
                <Plus
                  className="cursor-pointer"
                  size={40}
                  color="#ececf8"
                  onClick={() => handlePlusOrMinusClick("add")}
                />
                <input
                  className="flex size-12 items-center justify-center
                rounded-full border-2 border-accent-500 bg-background-100
                text-center text-2xl text-background-900"
                  type="text"
                  name=""
                  id=""
                  readOnly
                  value={productAmount}
                />
                <Minus
                  className="cursor-pointer"
                  size={40}
                  color="#ececf8"
                  onClick={() => handlePlusOrMinusClick("remove")}
                />
              </div>
              <button
                id="add-to-cart-btn"
                onClick={handleAddToCartClick}
                className="shadow-accent-btn flex h-14 items-center
                gap-4 rounded-md bg-primary-500 px-4 text-2xl
                font-medium shadow-primary-500 transition-shadow
                duration-100"
                type="submit"
              >
                <ShoppingCart size={32} />
                <p className="font-semibold">Add to Cart</p>
              </button>
            </div>
          </div>
        </section>
      </div>
    );
  }
};

export default Product;
