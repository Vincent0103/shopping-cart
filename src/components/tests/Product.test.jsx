import Loader from "../fetchUtils/Loader";
import ErrorShower from "../fetchUtils/ErrorShower";
import { render, screen, act } from "@testing-library/react";
import { afterEach, beforeEach, describe, it, expect, vi } from "vitest";
import Product from "../product/Product";
import Shop from "../shop/Shop";
import useData from "../fetchUtils/Fetch";
import {
  useParams,
  Link,
  createMemoryRouter,
  RouterProvider,
  Navigate,
} from "react-router-dom";
import userEvent from "@testing-library/user-event";

vi.mock("../product/Product.jsx", () => ({
  default: ({ mockedId, categoryName = "" }) => {
    const {
      data: productInfos,
      errorMessage,
      isLoading,
    } = useData(`https://fakestoreapi.com/products/${mockedId}`, [mockedId]);

    if (isLoading) return <Loader />;
    else if (errorMessage) return <ErrorShower errorMsg={errorMessage} />;
    else if (productInfos) {
      return (
        <>
          <Link to={`/shop/${categoryName}`}>
            <li>{categoryName}</li>
          </Link>
          <p>Rendered id: {productInfos.id}</p>
          <p>{productInfos.title}</p>
          <p>{productInfos.category}</p>
          <p>{productInfos.price}</p>
          <p>{productInfos.description}</p>
        </>
      );
    }
  },
}));

vi.mock("../shop/Shop.jsx", () => ({
  default: () => {
    let { name } = useParams();

    const validCategories = ["electronics", "mens-clothing"];

    name = name.split("/")[0];
    if (!validCategories.includes(name)) name = "";

    return (
      <>
        <nav>
          <ul>
            <Link>
              <li>
                <h3>{name}</h3>
              </li>
            </Link>
          </ul>
        </nav>
      </>
    );
  },
}));

describe("Shop Product", () => {
  describe("fetch", () => {
    beforeEach(() => {
      fetch = vi.fn();
    });

    afterEach(() => {
      vi.restoreAllMocks();
    });

    it("fetches the correct product when given a valid url with a valid id", async () => {
      fetch.mockResolvedValue({
        json: vi.fn().mockResolvedValue({
          id: 1,
          title: "gucci airpods",
          price: 85,
          category: "electronics",
          description: "airpods for luxury peoples",
          image: "...",
        }),
        ok: true,
      });

      const routes = [
        {
          path: "/",
          element: <Navigate to={"/products/:id"} replace />,
        },
        {
          path: "/products/:id",
          element: <Product mockedId={1} />,
        },
      ];
      const router = createMemoryRouter(routes);

      await act(async () => {
        render(<RouterProvider router={router} />);
      });
      expect(fetch).toHaveBeenCalled();

      const productTitle = screen.getByText(/gucci airpods/i);
      const productId = screen.getByText(/rendered id: 1/i);

      expect(productTitle).toBeInTheDocument();
      expect(productId).toBeInTheDocument();

      expect(screen).toMatchSnapshot();
    });

    it("shows an error message if the given id doesn't exist as a product", async () => {
      fetch.mockRejectedValue(new Error("Fetch failed"));

      await act(async () => {
        render(<Product />);
      });

      expect(screen.getByText(/error: fetch failed/i)).toBeInTheDocument();
    });
  });

  describe("history nav", () => {
    it("should link back to the original category when clicking the category link", async () => {
      fetch = vi.fn().mockResolvedValue({
        json: vi.fn().mockResolvedValue({
          id: 1,
          title: "gucci airpods",
          price: 85,
          category: "electronics",
          description: "airpods for luxury peoples",
          image: "...",
        }),
        ok: true,
      });

      const routes = [
        {
          path: "/",
          element: <Navigate to={"product/:id"} replace />,
        },
        {
          path: "/product/:id",
          element: <Product categoryName={"electronics"} />,
        },
        {
          path: "/shop?/:name",
          element: <Shop />,
        },
      ];

      const router = createMemoryRouter(routes);
      await act(async () => {
        render(<RouterProvider router={router} />);
      });

      const user = userEvent.setup();

      const categoryLink = screen.getByRole("link", { name: "electronics" });
      await user.click(categoryLink);

      // the mocked shop only displays a link with the given parameter from the history link name
      expect(
        screen.getByRole("heading", { name: "electronics" }),
      ).toBeInTheDocument();
    });
  });
});
