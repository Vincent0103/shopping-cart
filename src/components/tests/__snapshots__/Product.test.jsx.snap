// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop Product > fetch > fetches the correct product when given a valid url with a valid id 1`] = `
<div>
  <p>
    Rendered id: 
    1
  </p>
  <p>
    gucci airpods
  </p>
  <p>
    electronics
  </p>
  <p>
    85
  </p>
  <p>
    airpods for luxury peoples
  </p>
</div>
`;
